### ISSUE TITLE
Bug: [prefer-for-of] false positive when using non-null assertion within update expressions

### ISSUE THREAD

**Issue by zyoshoka** [typescript-eslint/typescript-eslint#10977]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=MYewdgzgLgBAhgJwXAnjAvDA2gRgDQwDMhBATACwEC0AnALoDcAUEwPSswgDWTAZiAhgAKADYBTWAEsMMAAwMY0gDzwkqAHTiwAcygALBZIDURgJSrkKLJLonmbDrziSRfAcPFSZ8xTBWJLTTEdfUMTcwDUazoAQjsmIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wlq4D2fAV3RRe0IZHBgAviFlA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const array = [1, 33, 24, -9];

// ok
for (let i = 0; i < array.length; i++) array[i]++;

// fail
for (let i = 0; i < array.length; i++) array[i]!++;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/prefer-for-of": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The 7th line shouldn't report `prefer-for-of` error.

### Actual Result

The 7th line reports `prefer-for-of` error.

**Comment by kirkwaiblinger** [comment]
> I have completed writing the patch and will wait until it is labeled [ accepting prs](https://github.com/typescript-eslint/typescript-eslint/labels/accepting%20prs) Go ahead, send a pull request that resolves this issue .

🚀
